conbench/api/compare.py

```diff
@@ -135,18 +136,36 @@ def __init__(
         threshold: Optional[float],
         threshold_z: Optional[float],
     ) -> None:
-        if (
-            baseline
-            and baseline.unit
-            and contender
-            and contender.unit
-            and baseline.unit != contender.unit
-        ):
-            raise UnmatchingUnitsError(
-                "Benchmark units do not match. Benchmark result with ID "
-                f"'{baseline.id}' has unit '{baseline.unit}' and benchmark result "
-                f"with ID '{contender.id}' has unit '{contender.unit}'."
-            )
+        # What do we know here? Is one of baseline and contender guaranteed
+        # to not be None?
+        assert any([baseline, contender])
+
+        # Can/should we assume that only non-errored results are added into a
+        # comparator? Then we can assume that the unit is set. Turns out: no,
+        # either might be errored. Thinking about possible code paths. Can
+        # `baseline` for example be an errored benchmark result while contender
+        # is `None`?
+
+        # `self.unit` is `None` for all cases except for when both contender
+        # and baseline are non-failed results and have the same unit; then it's
+        # that unit.
```

Comment:
This is specification I added here, I like the simplicity of that.

I couldn't quite see clearly how this was specified before.

Reply:
I think this is still a bit too restrictive and will cause visual confusion if we deploy. As I mentioned above, it's possible to submit a result with data and an error, and we have many examples in our current databases of that being the case. So I think the statement should be

>`self.unit` is `None` for all cases except for when both contender and baseline ~are non-failed results and~ have the same unit; then it's that unit.

Reply:
Ah, I had forgotten that this is what we settled on:
https://github.com/conbench/conbench/blob/fcb7e451c807a4c113cfcc9adc9bc4112ba5390e/conbench/entities/benchmark_result.py#L966-L968
So you're right that if either `baseline` or `contender` is failed, we shouldn't do any further analysis.

Reply:
> We could get rid of that and give None if there's an error.

Nice, yes. Adding a commit! I think let's do the relevant validity checks in the constructor, and then in the more mathy methods we can inspect the outcome of that validation.